### ISSUE TITLE
Input dropout incompatible with Model class.

### ISSUE THREAD

**Issue by nryant** [fchollet/keras#2335]
Currently, when applying dropout to an input ``x``, the ``_keras_history`` attribute of ``x`` is modified in place from ``keras.engine.topology.InputLayer`` to ``keras.layers.core.Dropout``:

    from keras.layers import Dropout, Input
    x = Input((800,), dtype='float32', name='input')
    print('Before: %r' % x._keras_history[0])
    x_drop = Dropout(0.5, name='input_drop')(x)
    print('After: %r' % x._keras_history[0])

which yields:

    Before: <keras.engine.topology.InputLayer object at 0x7ffb137dd310>
    After: <keras.engine.topology.InputLayer object at 0x7ffb137dd310>

When we then specify ``x`` as input to a new ``Model`` instance as in the following:

    from keras.layers import Activation
    from keras.models import Model
    y = Activation('linear')(x)
    model = Model(x, y)

there are two unintended consequences:

* a warning is generated to the effect that ``x`` is not an input tensor
* the ``input_names`` attribute of ``model`` is set incorrectly to ``['input_drop']`` rather than ``['input']``, so any call to ``train_on_batch``, ``fit_generator``, etc that specifies ``x`` as a dictionary will fail

This issue is specific to ``Dropout`` and does not arise when ``GaussianNoise`` or ``GaussianDropout`` is applied to an input.

-------

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

**Comment by fchollet** [comment]
There is no issue with Dropout with a first layer (or even as the only layer), i.e. with a model such as:
```python
from keras.layers import Dropout, Input
from keras.models import Model

x = Input(shape=(10,))
y = Dropout(0.5)(x)
model = Model(x, y)
```

**Comment by nryant** [comment]
Going back and looking at the original code that generated this, the problem arose because in the rather complicated model being built,  the default input dropout was 0, in which case ``Dropout.call`` just returns ``x``. Try:

    from keras.layers import Dropout, Input
    from keras.models import Model

    x = Input(shape=(10,))
    y = Dropout(0.0)(x)
    model = Model(x, y)

Now, the snippet I included when opening the ticket also leads to ``_keras_history`` being modified on ``x``, but that is completely coincidental and happens because I typoed and made ``x`` and not ``x_drop`` the input to the ``Activation`` layer, which just happened to have a linear activation function.
